iree/turbine/kernel/wave/layers/quant_linear.py

```diff
@@ -0,0 +1,283 @@
+    # This kernel uses the input sizes M, N, K throughout, as the tiling
+    # and data movement strategy is determined during the compilation process.
+    # These can be influenced by introducing constraints.
+    def gemm_core(a, b, c_reg, result):
+        # TODO: Registers for quantization scaling of inputs. Remove once scalar
+        # codegen is enabled.
+        a_scale = tkl.Register[B, M, K, tkl.f16](1 / input_scale.item())
+        a_clamp_max = tkl.Register[B, M, K, tkl.f16](qdtype_max)
+        a_clamp_min = tkl.Register[B, M, K, tkl.f16](qdtype_min)
+        b_scale = tkl.Register[N, K, tkl.f16](1 / weight_scale.item())
+        b_clamp_max = tkl.Register[N, K, tkl.f16](qdtype_max)
+        b_clamp_min = tkl.Register[N, K, tkl.f16](qdtype_min)
+        a_scale_deq = tkl.Register[B, M, N, tkl.f32](input_scale.item())
+        b_scale_deq = tkl.Register[B, M, N, tkl.f32](weight_scale.item())
+
+        @tkw.reduction(K, init_args=[c_reg])
+        def repeat(
+            acc: tkl.Register[B, M, N, tkl.f32],
+        ) -> tkl.Register[B, M, N, tkl.f32]:
+            a_reg = tkw.read(a)
+            b_reg = tkw.read(b)
+            a_reg *= a_scale
+            a_reg = clamp_tensor(a_reg, a_clamp_min, a_clamp_max)
+            b_reg *= b_scale
+            b_reg = clamp_tensor(b_reg, b_clamp_min, b_clamp_max)
+            acc = tkw.mma(a_reg, b_reg, acc)
+            acc *= a_scale_deq * b_scale_deq
+            return acc
+
+        tkw.write(
+            tkw.cast(repeat, tkl.f16),
+            result,
+        )
+
+    @tkw.wave(constraints)
+    def gemm(
+        a: tkl.Memory[B, M, K, ADDRESS_SPACE, tkl.f16],
+        b: tkl.Memory[N, K, ADDRESS_SPACE, tkl.f16],
+        result: tkl.Memory[B, M, N, GLOBAL_ADDRESS_SPACE, tkl.f16],
+    ):
+        c_reg = tkl.Register[B, M, N, tkl.f32](0.0)
+        gemm_core(a, b, c_reg, result)
+
+    @tkw.wave(constraints)
+    def gemm_with_bias(
+        a: tkl.Memory[B, M, K, ADDRESS_SPACE, tkl.f16],
+        b: tkl.Memory[N, K, ADDRESS_SPACE, tkl.f16],
+        bias: tkl.Memory[N, ADDRESS_SPACE, tkl.f16],
+        result: tkl.Memory[B, M, N, GLOBAL_ADDRESS_SPACE, tkl.f16],
+    ):
+        bias_reg = tkw.read(bias)
+        bias_reg = tkw.broadcast(bias_reg, target_shape=[B, M, N])
+        bias_reg = tkw.cast(bias_reg, tkl.f32)
+        # We can get "free" bias-add by setting bias as the initial
+        # value of accumulator to the mma
+        gemm_core(a, b, bias_reg, result)
+
+    hyperparams = {
+        ADDRESS_SPACE: SHARED_ADDRESS_SPACE,
+        LOAD_ELEMS_PER_THREAD: get_mfma_load_elems_per_thread(mfma_variant),
+        STORE_ELEMS_PER_THREAD: get_mfma_store_elems_per_thread(mfma_variant),
+        BLOCK_B: 1,
+        BLOCK_M: 64,
+        BLOCK_N: 64,
+        BLOCK_K: 32,
+        N: shape[1],
+        K: shape[0],
+    }
+    hyperparams.update(get_default_scheduling_params())
+
+    dynamic_symbols = [B, M]
```

Comment:
you will want to pass the actual values of B and M in the dynamic symbols map otherwise this won't work?

Reply:
Oh I see you are passing this below in the kernel.